1-Draft/RFCNNNN-Automatic-Resource-Management.md

```diff
@@ -0,0 +1,197 @@
+---
+RFC: RFC<four digit unique incrementing number assigned by Committee, this shall be left blank by the author>
+Author: Ilya Sazonov
+Status: Draft
+SupercededBy: <link to another RFC>
+Version: 1.0
+Area: Engine
+Comments Due: 7/1/2020
+Plan to implement: Maybe
+---
+
+# Automatic resource management
+
+## Rationale
+
+Currently PowerShell has limited capabilities compared with C# to excecute a mandatory code.
+
+PowerShell users can benefit from `finally` block of `try - catch` statement. It is very limited experience because resources in PowerShell can be distributed across scopes, cmdlets, pipelines and runspaces.
+
+If an user created a temporary file one should worry about deleting it, while PowerShell could register a mandatory code to delete this file and execute it automatically at the right time.
+
+PowerShell makes many smart things for users and could offer smart resource management.
+
+## Motivation
+
+As a script writer, I can use some well-known API-s which grabs resources
+and I want PowerShell understands this and releases the resources automatically at the right time
+so that avoids resorce leaks.
+
+As an advanced module and script writer, I can use some API-s which grabs resources
+and I want to register a code PowerShell must run to release the resources automatically at the right time so that avoids resorce leaks.
+
+As an advanced module and script writer, I want to make sure that a mandatory (cleanup, logging) code is always executed even if a pipeline / script / cmdlet is unexpectedly interrupted.
+
+As binary module developer I want to benefit from this in binary modules too.
+
+## User Experience
+
+User has no needs explicitly dispose an object when out of the script block:
+
+```powershell
+# some script commands
+...
+    # new script block
+    {
+        # PowerShell detects and registers new IDisposable object for the script block scope
+        $fs = [System.IO.File]::Open("file", [System.IO.FileMode]::Open)
+        ...
+        # Here the script block is closed and
+        # PowerShell Engine calls implicitly $fs.Dispose()
+        # before destroying the scope.
```

Comment:
How could PowerShell Engine know whether or not to call `$fs.Dispose()`? What if there was a call, `[Some.DotNetType]::Method( $fs )`... how does PS know if `$fs` is still in use or not?

Reply:
One that I know! :-)

> To help ensure that resources are always cleaned up appropriately, a Dispose method should be idempotent, such that it is callable multiple times without throwing an exception. Furthermore, subsequent invocations of Dispose should do nothing.

From https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose

Reply:
> How could PowerShell Engine know whether or not to call $fs.Dispose()? 

This is a really good point. Even if PowerShell could *prove* that all of its own references to `$fs` were no longer in use (by doing something like crawling the variable stack), it wouldn't be able to see if there were remaining references to `$fs` beyond the veil of .NET; PowerShell doesn't own the whole runtime. The only thing that has that kind of power is the GC, and that's why you implement a finalizer for unmanaged resources. As it happens, [FileStream does this already](https://github.com/dotnet/runtime/blob/110282c71b3f7e1f91ea339953f4a0eba362a62c/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L845-L851).